core/src/main/scalajvm/scalapb/zio_grpc/server/ZServerCallHandler.scala

```diff
@@ -108,30 +110,61 @@ object ZServerCallHandler {
       call: ZServerCall[Res],
       stream: ZStream[Any, Status, Res]
   ): ZIO[Any, Status, Unit] = {
-    def innerLoop(queue: Dequeue[Exit[Option[Status], Res]]): ZIO[Any, Status, Boolean] =
-      queue.take
-        .flatMap {
-          case Exit.Success(res)   => call.sendMessage(res).as(true)
-          case Exit.Failure(cause) =>
-            cause.failureOrCause match {
-              case Left(Some(status)) => ZIO.fail(status)
-              case Left(None)         => ZIO.succeed(false)
-              case Right(cause)       => ZIO.failCause(cause)
-            }
-        }
-        .repeatWhileZIO(res => call.isReady.map(_ && res))
-
-    def outerLoop(queue: Dequeue[Exit[Option[Status], Res]]): ZIO[Any, Status, Boolean] =
-      (call.awaitReady *> innerLoop(queue))
-        .repeatWhile(identity)
+    def takeFromQueue(queue: Dequeue[Exit[Option[Status], Res]]): ZIO[Any, Status, Unit] =
+      queue.takeAll.flatMap(takeFromCache(_, queue))
+
+    def takeFromCache(
+        xs: Chunk[Exit[Option[Status], Res]],
+        queue: Dequeue[Exit[Option[Status], Res]]
+    ): ZIO[Any, Status, Unit] =
+      ZIO.succeed {
+        var i                          = 0
```

Comment:
Maybe tail recursion would help keeping this more maintainable and avoid the imperative style?

Reply:
Ouch, i can try but am not sure it will make things clearer. Do you prefer this?
```
@tailrec def innerLoop(i: Int): IO[Status, Unit] =
          if (i < xs.length) {
            xs(i) match {
              case Failure(cause) =>
                cause.failureOrCause match {
                  case Left(Some(status)) =>
                    ZIO.fail(status)
                  case Left(None)         =>
                    ZIO.unit
                  case Right(cause)       =>
                    ZIO.failCause(cause)
                }
              case Success(value) =>
                call.call.sendMessage(value)
                if (call.call.isReady)
                  innerLoop(i + 1)
                else if (i == xs.length - 1)
                  takeFromQueue(queue)
                else
                  outerLoop(Some(xs.drop(i + 1)), queue)
            }
          } else takeFromQueue(queue)

        innerLoop(0)
```

Reply:
Thanks for rewriting. Yes, I'd prefer the tail-recursive version. Can you add some comments in the code, quickly looking at it here are a few questions that pop up:
* when processing the i-th element in the `Success` case, why do we drop `i+1` ?
* Is call.call.sendMessage a side-effect that we're calling directly from here?
* `else if (i == xs.length - 1)` isn't that case covered by the exit condition (top-level if..else)?

Reply:
> Ouch

Sorry if my comment came across as a criticism, that wasn't the intent - it's more about what style I personally find  easier to work with. Very appreciative of all your contributions!

Reply:
> Sorry if my comment came across as a criticism,

Not at all :-) No problem les let me work this out.

Reply:
Done, let me know what you think 🙏  This shaves 1s off the local benchmark, so hopefully this will be a nice boost in real conditions.